Update queue workflow in Farm to improve performance and clearer

lamhieu-vk opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

I think the current Farm queue workflow in Jest Worker  is quite complex and I think improving it makes the source code simpler and easier to understand by using native language operations, I think in addition to being easier to understand, it also improves a bit performance.

## Comparison (Updated 1)

### Run `loadTest` by script:
```
$ node --expose-gc src/__performance_tests__/test.js loadTest 100000
```

Result:
```
# Master branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 110747, wFPT: 110593 }
total jest-worker: { jWGT: 97244, jWPT: 97172 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 12.192655331521395
% improvement over 100000 calls (processing time): 12.135487779515882

# This branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 96077, wFPT: 95918 }
total jest-worker: { jWGT: 67547, jWPT: 67477 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 29.694932189806092
% improvement over 100000 calls (processing time): 29.651368877582936
```

### Run `empty` by script:
```
$ node --expose-gc src/__performance_tests__/test.js empty 100000
```

Result:
```
# Master branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 51264, wFPT: 51129 }
total jest-worker: { jWGT: 12657, jWPT: 12587 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 75.31015917602996
% improvement over 100000 calls (processing time): 75.38187721254083

# This branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 51633, wFPT: 51482 }
total jest-worker: { jWGT: 15611, jWPT: 15537 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 69.76546007398369
% improvement over 100000 calls (processing time): 69.82051979332583
```

**Comment**

- `loadTest`: **this branch** have a result greater than **master branch** - (29.69 vs 12.19)
- `empty`:  **this branch** have a result smaller than **master branch** - (69.76 vs 75.31)

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test plan

Green CI, make all tests are passed as the previous version and write more tests for new code.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

drew-gross commented:
# [Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=h1) Report
> Merging [#9521](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=desc) into [master](https://codecov.io/gh/facebook/jest/commit/c4ae594d1f09fe37542da0597ad979dacdf27345&el=desc) will **decrease** coverage by `0.00%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/facebook/jest/pull/9521/graphs/tree.svg?width=650&height=150&src=pr&token=oaWHH63dBr)](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #9521      +/-   ##
==========================================
- Coverage   65.03%   65.03%   -0.01%     
==========================================
  Files         283      283              
  Lines       12158    12154       -4     
  Branches     3014     3012       -2     
==========================================
- Hits         7907     7904       -3     
+ Misses       3608     3607       -1     
  Partials      643      643              
```


| [Impacted Files](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [packages/expect/src/utils.ts](https://codecov.io/gh/facebook/jest/pull/9521/diff?src=pr&el=tree#diff-cGFja2FnZXMvZXhwZWN0L3NyYy91dGlscy50cw==) | `96.22% <0.00%> (+1.25%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=footer). Last update [c4ae594...018b959](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

SimenB commented on CHANGELOG.md:
```diff
@@ -461,6 +461,7 @@ We skipped 24.2.0 because a draft was accidentally published. Please use `24.3.0
 - `[jest-worker]`: Migrate to TypeScript ([#7853](https://github.com/facebook/jest/pull/7853))
 - `[jest]`: Migrate to TypeScript ([#8024](https://github.com/facebook/jest/pull/8024))
 - `[pretty-format]`: Migrate to TypeScript ([#7809](https://github.com/facebook/jest/pull/7809), [#7809](https://github.com/facebook/jest/pull/7972))
+- `[jest-worker]`: Update queue workflow in Farm to improve performance and clearer ([#9521](https://github.com/facebook/jest/pull/9521)
```
should be under `performance` at the top of the file (under master)

lamhieu-vk replied:
I updated it

SimenB commented:
Super cool, thanks @lamhieu-vk! There are some benchmarks in `src/__performance_tests__`, could you run them with and without your change?

lamhieu-vk commented on packages/jest-worker/src/base/BaseWorkerPool.ts:
```diff
@@ -112,11 +112,9 @@ export default class BaseWorkerPool {
-      }),
-      {forceExited: false},
-    );
+    const forceExited = workerExits.some(isExited => isExited);
```
I updated it, thanks!

lamhieu-vk commented:
I updated PR to show the results of performance tests and split queue to class.

SimenB commented on packages/jest-worker/src/Queue.ts:
```diff
@@ -0,0 +1,36 @@
+    this.offset = 0;
+  }
+
+  public push = (item: QueueChildMessage) => this.data.push(item);
```
```suggestion
  push(item: QueueChildMessage) {
    this.data.push(item)
  }
```

somewhat same for the others - no need for the `public` modifier and don't use arrow functions

lamhieu-vk replied:
I will update it now

SimenB commented on packages/jest-worker/src/__tests__/Queue.test.js:
```diff
@@ -0,0 +1,83 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+'use strict';
```
no need for `user strict` when using `import` - it's implicit

lamhieu-vk replied:
I cloned this template from `Farm.test.js` but I will remove this line

lamhieu-vk commented:
I updated and green is back!

thymikee commented:
So... based on your perf test results we ended up with more code and worse performance overall?

thymikee commented:
My local `loadTest` on smaller set of 10000:
```
master
% improvement over 10000 calls (global time): 4.8286937901498925
% improvement over 10000 calls (processing time): 5.373166994966076

this branch
% improvement over 10000 calls (global time): 9.502121933547253
% improvement over 10000 calls (processing time): 10.185771585391597
```
and 1000:
```
master
% improvement over 1000 calls (global time): 2.2794117647058822
% improvement over 1000 calls (processing time): 4.366438356164384

this branch
% improvement over 1000 calls (global time): -3.3280507131537242
% improvement over 1000 calls (processing time): -1.20817843866171
```

And OP tests for 100000 show quite a significant drop (9.85 vs 20.34 on master).

SimenB commented:
Hah, I'm confused by the perf test. I'm used to them measuring how many operations per second they do (so higher is better), but these seem to measure time to complete x amount of operations instead (so lower is better). Which is it? 😬

thymikee commented:
This shows the % of difference between worker-farm and jest-worker. The higher on positive scale, the better

lamhieu-vk commented:
I don't know which values to use as my main goal, can someone tell me which values I need to care more about? 😳

lamhieu-vk commented:
I updated PR and result, I care about `loadTest` results than `empty` test!

drew-gross commented:
# [Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=h1) Report
> Merging [#9521](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=desc) into [master](https://codecov.io/gh/facebook/jest/commit/c4ae594d1f09fe37542da0597ad979dacdf27345&el=desc) will **increase** coverage by `0.00%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/facebook/jest/pull/9521/graphs/tree.svg?width=650&height=150&src=pr&token=oaWHH63dBr)](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master    #9521   +/-   ##
=======================================
  Coverage   65.03%   65.04%           
=======================================
  Files         283      284    +1     
  Lines       12158    12161    +3     
  Branches     3014     3012    -2     
=======================================
+ Hits         7907     7910    +3     
  Misses       3608     3608           
  Partials      643      643           
```


| [Impacted Files](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [packages/jest-worker/src/Queue.ts](https://codecov.io/gh/facebook/jest/pull/9521/diff?src=pr&el=tree#diff-cGFja2FnZXMvamVzdC13b3JrZXIvc3JjL1F1ZXVlLnRz) | `100.00% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=footer). Last update [c4ae594...a748372](https://codecov.io/gh/facebook/jest/pull/9521?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

lamhieu-vk commented:
I updated result, please give me a feedback!

cpojer commented:
I do not understand the intention behind this PR. The top post and all comments imply that performance with this change will be worse. Performance of `jest-worker` is critical to Jest and many other tools, merging code that regresses it for readability purposes is not ok.

lamhieu-vk commented:
@cpjoer I updated an it faster old queue on loadTest (2.5x)

lamhieu-vk commented:
Please review newest result with new queue.

### Run `loadTest` by script:
```
$ node --expose-gc src/__performance_tests__/test.js loadTest 100000
```

Result:
```
# Master branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 110747, wFPT: 110593 }
total jest-worker: { jWGT: 97244, jWPT: 97172 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 12.192655331521395
% improvement over 100000 calls (processing time): 12.135487779515882

# This branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 96077, wFPT: 95918 }
total jest-worker: { jWGT: 67547, jWPT: 67477 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 29.694932189806092
% improvement over 100000 calls (processing time): 29.651368877582936
```

### Run `empty` by script:
```
$ node --expose-gc src/__performance_tests__/test.js empty 100000
```

Result:
```
# Master branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 51264, wFPT: 51129 }
total jest-worker: { jWGT: 12657, jWPT: 12587 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 75.31015917602996
% improvement over 100000 calls (processing time): 75.38187721254083

# This branch
---------------------------------------------------------------------------
total worker-farm: { wFGT: 51633, wFPT: 51482 }
total jest-worker: { jWGT: 15611, jWPT: 15537 }
---------------------------------------------------------------------------
% improvement over 100000 calls (global time): 69.76546007398369
% improvement over 100000 calls (processing time): 69.82051979332583
```

**Comment**

- `loadTest`: **this branch** have a result greater than **master branch** - (29.69 vs 12.19)
- `empty`:  **this branch** have a result smaller than **master branch** - (69.76 vs 75.31)

cpojer commented:
Sorry I still don't understand your results. It seems like this is 3x slower in the average case (loadTest)?

lamhieu-vk commented:
nope, the latest results I run test performance with loop is 4, default is 10. you can try it!

cpojer commented:
Can you clarify what the concrete performance difference is with your diff applied?

lamhieu-vk commented:
After compare many way, I think (after run performance test on jest-worker):
set > linkedlist > native array.

lamhieu-vk commented on packages/jest-worker/src/Queue.ts:
```diff
@@ -0,0 +1,33 @@
+
+  constructor() {
+    this.data = new Set();
+    this.iterator = this.data.values();
```
@cpojer I implemented new queue base Set, it faster linkedlist we using now and clearer. Bẻoe, I tries to implemented with native Array but it is slower.

thymikee commented:
I think the confusing part in these perf tests is this bit, talking about _percentage_ improvement:
```
% improvement over 100000 calls (processing time): 69.82051979332583
```
It would be clearer with this:
```
Over 100000 calls (processing time) jest-worker is faster by: 69.82%
```

lamhieu-vk commented:
@thymikee would you want to me update it? text log
of performance test.

lamhieu-vk commented:
sorry about my comments because I reply on my phone :(

lamhieu-vk commented:
I think we confusing in the PR with many comments above and I want to summary last updated:
- I updated the performance test report to clearer.
- I updated Queue in Farm, using `Set iterator` instead of `LinkedList` to clearer and optimal performance (see performance test results below). Run `loadTest` over 100000 calls.

```js
## LinkedList (master branch)
---------------------------------------------------------------------------
total worker-farm: { globalTime: 259424, processingTime: 258953 }
total jest-worker: { globalTime: 196822, processingTime: 196562 }
---------------------------------------------------------------------------
Over 100000 calls (global time) jest-worker is faster than worker-farm by: 24.13115209078574%
Over 100000 calls (processing time) jest-worker is faster than worker-farm by: 24.0935613798643%

## Set (this branch)
---------------------------------------------------------------------------
total worker-farm: { globalTime: 303149, processingTime: 302400 }
total jest-worker: { globalTime: 192675, processingTime: 192436 }
---------------------------------------------------------------------------
Over 100000 calls (global time) jest-worker is faster than worker-farm by: 36.44214561156395%
Over 100000 calls (processing time) jest-worker is faster than worker-farm by: 36.363756613756614%
```

lamhieu-vk commented:
@jeysal would you be able to take a quick look? I updated it.

jeysal commented:
Would love to see what this evolved into (I haven't looked since the comment where I ran the benchmark) but not sure when I'll have time to get to it.

lamhieu-vk commented:
thanks @jeysal, I waiting for any feedback!

lamhieu-vk commented:
hi, I just wanted to say I'm still waiting for something here 😑😚.

jeysal commented:
Hi, I know, sorry, my inbox currently goes back to Feb 6 :|

lamhieu-vk commented:
I know and github is changing it 🎉. Please make sure you pulled new code in this branch to review and run performance test, thanks 😁

jeysal commented:
I finally managed to take a look at the implementation and it does look correct to me. As JS Sets are linked sets, it uses essentially the same data structure as well. But because of that in your benchmark numbers it seems to perform the same as `master` as well. I'm not sure why your `worker-farm` numbers changed by branch, making it look like a difference even though `jest-worker` remained the same. When I ran it both `worker-farm` and `jest-worker` were essentially unchanged between branches.

lamhieu-vk commented:
While I was working with Set and Map, I found that its performance was always higher than Array and Object for tasks like this. I did it right by running the load test and it was no surprise that it gave a better number.
Do you want me to do something else?

jeysal commented:
Hey, check the numbers you posted again - they don't actually show a better number for `jest-worker` with your changes vs master, they for some reason show a worse number for `worker-farm` (which is the worker package from npm used as the baseline) on your branch.

cpojer commented:
Thanks for the check @jeysal. I think we can close this PR.
